pkg/gpu/cuda/cubin.go

```diff
@@ -0,0 +1,281 @@
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+// Generate String() methods for nvInfoAttr,nvInfoFormat enums so they can be printed in logs/error messages
+//go:generate go run golang.org/x/tools/cmd/stringer@latest -output cubin_string.go -type=nvInfoAttr,nvInfoFormat -linecomment
+
+package cuda
+
+import (
+	"bytes"
+	"debug/elf"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+)
+
+// nvInfoFormat defines the format (size) of a value of a .nv.info item
+type nvInfoFormat uint8
+
+const (
+	nviFmtNone nvInfoFormat = 0x01
+	nviFmtBval nvInfoFormat = 0x02
+	nviFmtHval nvInfoFormat = 0x03
+	nviFmtSval nvInfoFormat = 0x04
+)
+
+// nvInfoAttr defines the attribute of a .nv.info item
+type nvInfoAttr uint8
+
+const (
+	nviAttrError nvInfoAttr = iota
+	nviAttrPad
+	nviAttrImageSlot
+	nviAttrJumptableRelocs
+	nviAttrCtaidzUsed
+	nviAttrMaxThreads
+	nviAttrImageOffset
+	nviAttrImageSize
+	nviAttrTextureNormalized
+	nviAttrSamplerInit
+	nviAttrParamCbank
+	nviAttrSmemParamOffsets
+	nviAttrCbankParamOffsets
+	nviAttrSyncStack
+	nviAttrTexidSampidMap
+	nviAttrExterns
+	nviAttrReqntid
+	nviAttrFrameSize
+	nviAttrMinStackSize
+	nviAttrSamplerForceUnnormalized
+	nviAttrBindlessImageOffsets
+	nviAttrBindlessTextureBank
+	nviAttrBindlessSurfaceBank
+	nviAttrKparamInfo
+	nviAttrSmemParamSize
+	nviAttrCbankParamSize
+	nviAttrQueryNumattrib
+	nviAttrMaxregCount
+	nviAttrExitInstrOffsets
+	nviAttrS2rctaidInstrOffsets
+	nviAttrCrsStackSize
+	nviAttrNeedCnpWrapper
+	nviAttrNeedCnpPatch
+	nviAttrExplicitCaching
+	nviAttrIstypepUsed
+	nviAttrMaxStackSize
+	nviAttrSuqUsed
+	nviAttrLdCachemodInstrOffsets
+	nviAttrLoadCacheRequest
+	nviAttrAtomSysInstrOffsets
+	nviAttrCoopGroupInstrOffsets
+	nviAttrCoopGroupMaxRegids
+	nviAttrSw1850030War
+	nviAttrWmmaUsed
+	nviAttrHasPreV10Object
+	nviAttrAtomf16EmulInstrOffsets
+	nviAttrAtom16EmulInstrRegMap
+	nviAttrRegcount
+	nviAttrSw2393858War
+	nviAttrIntWarpWideInstrOffsets
+	nviAttrSharedScratch
+	nviAttrStatistics
+
+	// New between cuda 10.2 and 11.6
+	nviAttrIndirectBranchTargets
+	nviAttrSw2861232War
+	nviAttrSwWar
+	nviAttrCudaApiVersion
+	nviAttrNumMbarriers
+	nviAttrMbarrierInstrOffsets
+	nviAttrCoroutineResumeIdOffsets
+	nviAttrSamRegionStackSize
+	nviAttrPerRegTargetPerfStats
+
+	// New between cuda 11.6 and 11.8
+	nviAttrCtaPerCluster
+	nviAttrExplicitCluster
+	nviAttrMaxClusterRank
+	nviAttrInstrRegMap
+)
+
+// nvInfoItem is the in-file representation of an item  header in the .nv.info section. The value follows
+// according to the format
+type nvInfoItem struct {
+	Format nvInfoFormat
+	Attr   nvInfoAttr
+}
+
+// cubinParser is a helper struct to parse the cubin ELF sections
+type cubinParser struct {
+	kernels            map[string]*CubinKernel
+	sectPrefixToParser map[string]func(*elf.Section, *bytes.Reader, string) error
```

Comment:
consider renaming to `sectionParsers`

Reply:
In this case I prefer to make it explicit that the keys are section prefixes, not just section names.